### ISSUE TITLE
Adding callback to MapviewWidget that executes after the viewmodel is created

### ISSUE THREAD

**Issue by vasilevzhivko** [mikes222/mapsforge_flutter#104]
I want to start positioning service as soon as the app loads.
This was working when I was using FlutterMapView instead of MapviewWidget but with MapviewWidget the creation of the viewmodel and mapmodel are now Future based and when I try to start positioning from my map view provider I get an error that the viewModel is not initialised yet.

I was thinking if it makes sense to introduce something like a callback that can be executed as soon as the view model is created.
Something like that:

```
class MapviewWidget extends StatefulWidget {
  final DisplayModel displayModel;
  final CreateMapModel createMapModel;
  final CreateViewModel createViewModel;
  final VoidCallback onViewModelCreated; // Add this line
  final String? changeKey;

  const MapviewWidget({
    Key? key,
    required this.displayModel,
    required this.createMapModel,
    required this.createViewModel,
    required this.onViewModelCreated, // Add this line
    this.changeKey,
  }) : super(key: key);

  @override
  State<StatefulWidget> createState() {
    return _MapviewWidgetState();
  }
}
```

Then:
```
Widget _createViewModel(Func child) {
  if (_viewModel != null) return child();
  return FutureBuilder<ViewModel>(
      future: widget.createViewModel(),
      builder: (BuildContext context, AsyncSnapshot<ViewModel> snapshot) {
        if (snapshot.connectionState == ConnectionState.waiting)
          return progress("Waiting for View");
        if (snapshot.hasError) {
          _log.warning(snapshot.stackTrace);
          return Text(
              "Error while creating view: ${snapshot.error?.toString()}",
              style: const TextStyle(color: Colors.red));
        }
        if (snapshot.data == null) return progress("Creating View");
        _viewModel = snapshot.data;
        _subscription = _viewModel!.observePosition
            .listen((MapViewPosition mapViewPosition) async {
          if (mapViewPosition.hasPosition())
            _jobQueue!.submitJobSet(_viewModel!, mapViewPosition);
        });
        widget.onViewModelCreated(); // Add this line
        _log.info(
            "ViewModel created in connectionState ${snapshot.connectionState.toString()}");
        return child();
      });
}
```

in that case when I use the map I can do the following:

```
MapviewWidget(
  displayModel: mapViewModel.displayModel,
  createMapModel: mapViewModel.createMapModel,
  createViewModel: mapViewModel.createViewModel,
  onViewModelCreated: () {
    mapViewModel.startGpsListening(); // Add this line
  },
),
```

**Comment by mikes222** [comment]
The ``createViewModel()`` is already a callback method and implemented by your own code. You can do anything you want there. 

You have 2 different domains. One domain (=set of classes) for gps listening and another domain (=mapsforge) for viewing the map. 
Life will be easier if you do not tight coupling these two domains together. Instead use streams or similar decoupling techniques. You need a manager or controller or however you may call it to manage these two domains and decouple them. 

In other words:
- Whenever you create or dispose the mapsforge view send events. 
- Whenever gps has new coordinates send events. 

So your gps domain can listen if gps is necessary (if the map has been opened) and the manager/controller can listen to positioning events and update the map. 

I doubt that your error is really because the viewModel is not initialized. Unfortunately I do not have a stacktrace to find the root cause.

**Comment by vasilevzhivko** [comment]
Got it, thanks!
